Revise Since RFC. (#226)
nervosnetwork/rfcs: rfcs/0017-tx-valid-since/0017-tx-valid-since.md

---
Number: "0017"
Category: Standards Track
Status: Active
Author: Jinyang Jiang <@jjyr>, Ian Yang <@doitian>
Created: 2019-03-11
---

# Transaction Since Precondition

<!-- Diagrams are created in Lucid: https://lucid.app/documents/view/d756089a-2388-4ea4-b61a-3943cbe2620a -->

## Abstract

This RFC suggests adding a consensus rule to restrict committing a transaction before a specified time. The time comes from the block headers in the chain via block number, epoch number with fraction, or block timestamp.

## Summary

The new consensus rule allows the transaction input to specify an optional since precondition. CKB nodes must verify the transactions in the commitment zone of a block that all the input since preconditions are fulfilled.

> The since precondition locates a unique block in any chain if it is long enough. The precondition is effective before the block, and is fulfilled since the block.

Reviewer: ```suggestion
The since precondition locates a unique block in the past or future on the chain. The precondition is effective before the block, and is fulfilled since the block.
```

There are three metrics to specify the since precondition:

1. via block number,
2. epoch number with fraction,
3. or the median timestamp of the preceding 37 blocks.

All of them strictly increase along the block number.

> The precondition is absolute or relative. A relative precondition depends on which block has committed the input.

Reviewer: "which block has committed the input" sounds ambiguous - it could mean the block commits the input, or the block commits the cell pointed by the input.

> The precondition is absolute or relative. A relative precondition depends on which block has committed the input.

Author: How about this?

```suggestion
The precondition is absolute or relative. A relative precondition depends on which block has committed the cell referenced by the input.
```

![](cell-lifecycle.jpg)

## Specification

### Encoding

The `since` is a `u64` (unsigned 64-bit integer) type field in the transaction input. The highest 8 bits of the `since` field is the `flags`. The remaining `56` bits represent the `value`.

![](since-encoding.jpg)

* The highest bit is the relative flag.
    * `0`: The `value` is absolute that the cell destruction time should not be earlier than `value`.
    * `1`: The `value` is relative that the cell destruction time should not be earlier than cell creation time plus `value`.
* The following two bits choose the time metric used to represent the creation time, destruction time, and the `value`.
    * `00`: Use block number.
    * `01`: Use epoch number with fraction.
    * `10`: Use the timestamp median of previous block headers.
    * `11`: Invalid. Transaction should not set metric flag to `11`.
* The next 5 bits are reserved for future extension. They must be set to all zeros now.

### Since Verification

> The consensus iterates each input to check whether the cell is mature:

Reviewer: ```suggestion
A CKB node must iterate each input to check whether the cell is mature:
```

Another problem is "the cell is mature" is confusing/improper, "mature" in this RFC is actually an attribute of input not cell.

> * The cell is mature when the `since` is 0.

Reviewer: ```suggestion
* An input is always due if its `since` is 0.
```

An example if we use the word "due" with "input".

> * If the relative flag is `0` (absolute), the cell is mature when the destruction time is larger than or equal to the `value`.

Reviewer: I think using "destruction time" to define validation rules is inappropriate, because "destruction time" can only be determined when the transaction is included in a block, but the validation here happens before the transaction is included.

Number: "0017"
Category: Standards Track
Status: Proposal
Author: Jinyang Jiang, Ian Yang
Organization: Nervos Foundation
Created: 2019-03-11
---

# Transaction valid since

> <!-- Dialogs are created in LucidChart: https://lucid.app/documents/view/d756089a-2388-4ea4-b61a-3943cbe2620a -->

Author: ```suggestion
<!-- Diagrams are created in LucidChart: https://lucid.app/documents/view/d756089a-2388-4ea4-b61a-3943cbe2620a -->
```

    * `E(S)` is the epoch number of the epoch that the block `S` is in.
    * `I(S)` is the block index of `S` in the epoch. The index of the first block in an epoch is 0.
    * `L(S)` is the epoch length.
* The destruction time is a rational number `E(T) + I(T) / L(T)`, where
    * `E(T)` is the epoch number of the epoch that the block `T` is in.
    * `I(T)` is the block index of `T` in its epoch.
    * `L(T)` is the epoch length.
* The `value` is a encoded rational number `E + I / L` the same as the `epoch` field in the block header, where

    ![](../0027-block-structure/epoch.png)

    * `E` is the epoch number part, the three bytes from the lowest bit 0 to 23.
    * `I` is the block index in the epoch part, the two bytes from the bit 24 to 39.
    * `L` is the epoch length part, the two bytes from the bit 40 to 55. There's an **exception** here, `L` is at least 1. If `L` is zero, the whole rational number is considered as `E + 0 / 1`.

**Specially**, when the relative flag is `0` (absolute), the `value` is normalized first. But when the relative flag is `1` (relative), the `value` is interpreted as it is and is added to the creation time directly.

> **Absolute Epoch With Fraction Value Normalization**: If the relative flag is `0` (absolute)　and `I` is larger than or equal to `L`, the `value` is normalized to `E + 1 + 0 / 1`.

Reviewer 2: So, by following this rule, we normalize `100 + 201 / 100` to `101 + 0 / 1`. IMO it is a counter-intuitive behaviour, why can't we just reject these values?

Reviewer 2: This phase is outdated #233

Author: Yes, deprecated after #223 is activated. A notice should be added that the rule is removed after #223 is merged. Before that, thia PR should be reviewed based on the current consensus edition.

Only after #223 is activated in testnet and mainnet, and there are no history blocks in testnet and mainnet violating #223, can we remove it from this RFC.
